docs/source/format/CDataInterface.rst

```diff
@@ -0,0 +1,773 @@
++------------------------+---------------------------------------------------+------------+
+| ``+s``                 | struct                                            |            |
++------------------------+---------------------------------------------------+------------+
+| ``+m``                 | map                                               | \(2)       |
++------------------------+---------------------------------------------------+------------+
+| ``+ud:I,J,...``        | dense union with type ids I,J...                  |            |
++------------------------+---------------------------------------------------+------------+
+| ``+us:I,J,...``        | sparse union with type ids I,J...                 |            |
++------------------------+---------------------------------------------------+------------+
+
+Notes:
+
+(1)
+   The timezone string is appended as-is after the colon character ``:``, without
+   any quotes.
+
+(2)
+   As specified in the Arrow columnar format, the map type has a single child type
+   named ``entries``, itself a 2-child struct type of ``(key, value)``.
+
+Examples
+--------
+
+* A dictionary-encoded ``decimal128(precision = 12, scale = 5)`` array
+  with ``int16`` indices has format string ``s``, and its dependent dictionary
+  array has format string ``d:12,5``.
+* A ``list<uint64>`` array has format string ``+l``, and its single child
+  has format string ``L``.
+* A ``struct<ints: int32, floats: float32>`` has format string ``+s``; its two
+  children have names ``ints`` and ``floats``, and format strings ``i`` and
+  ``f`` respectively.
+* A ``map<string, float64>`` array has format string ``+m``; its single child
+  has name ``entries`` and format string ``+s``; its two grandchildren have names
+  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
+* A ``sparse_union<ints: int32, floats: float32>`` with type ids ``4, 5``
+  has format string ``+us:4,5``; its two children have names ``ints`` and
+  ``floats``, and format strings ``i`` and ``f`` respectively.
+
+
+Structure definitions
+=====================
+
+The following free-standing definitions are enough to support the Arrow
+C data interface in your project.  Like the rest of the Arrow project, they
+are available under the Apache License 2.0.
+
+.. code-block:: c
+
+   #define ARROW_FLAG_ORDERED 1
+   #define ARROW_FLAG_NULLABLE 2
+   #define ARROW_FLAG_KEYS_SORTED 4
+
+   struct ArrowArray {
+     // Type description
+     const char* format;
+     const char* name;
+     const char* metadata;
+     int64_t flags;
+
+     // Data description
+     int64_t length;
+     int64_t null_count;
+     int64_t offset;
+     int64_t n_buffers;
+     int64_t n_children;
+     const void** buffers;
+     struct ArrowArray** children;
+     struct ArrowArray* dictionary;
+
+     // Release callback
+     void (*release)(struct ArrowArray*);
```

Comment:
One question that just occurred to me. Does it make sense for `release` to return 0 on success and other value on value? Or should the expectation be as it currently is that `release` cannot fail

Reply:
It's a destructor, so it should not fail to release resources.

Reply:
OK